open_seq2seq/data/text2speech/text2speech.py

```diff
@@ -134,10 +134,66 @@ def __init__(self, params, model, num_workers=None, worker_id=None):
     self.params['idx2char'] = {i: w for w, i in self.params['char2idx'].items()}
     self.params['src_vocab_size'] = len(self.params['char2idx'])
 
-    n_mels = self.params['num_audio_features']
+    n_feats = self.params['num_audio_features']
+    if "both" in self.params["output_type"]:
+      self._both = True
+      if self.params["feature_normalize"]:
+        raise ValueError(
+            "feature normalize is not currently enabled for both mode"
+        )
+      if not isinstance(n_feats, dict):
+        raise ValueError(
+            "num_audio_features must be a dictionary for both mode"
+        )
+      else:
+        if ("mel" not in n_feats and
+            "magnitude" not in n_feats):
+          raise ValueError(
+            "num_audio_features must contain mel and magnitude keys"
+          )
+        elif (not isinstance(n_feats["mel"], int) or
+              not isinstance(n_feats["magnitude"], int)):
+            raise ValueError(
+                "num_audio_features must be a int"
+            )
+      n_mels = n_feats['mel']
+      data_min = self.params.get("data_min", None)
+      if data_min is not None:
+        if not isinstance(data_min, dict):
+          raise ValueError(
+              "data_min must be a dictionary for both mode"
+          )
+        else:
+          if "mel" not in data_min and "magnitude" not in data_min:
+            raise ValueError(
+              "data_min must contain mel and magnitude keys"
+            )
+          elif (not isinstance(data_min["mel"], float) or 
+                not isinstance(data_min["magnitude"], float)):
+            raise ValueError(
+                "data_min must be a float"
+            )
+      self._exp_mag = self.params.get("exp_mag", True)
+    else:
+      if not isinstance(n_feats, int):
+        raise ValueError(
+            "num_audio_features must be a float for mel or magnitude mode"
+        )
+      if not isinstance(self.params.get("data_min",1.0), float):
+        raise ValueError(
+            "data_min must be a float for mel or magnitude mode"
+        )
+      self._both = False
+      self._exp_mag = False
+      n_mels = n_feats
+
 
     if "mel" in self.params["output_type"]:
```

Comment:
Slightly shorter and cleaner version:
self._mel = "mel" in self.params["output_type"]